senders/opsgenie/send.go

```diff
@@ -0,0 +1,152 @@
+package opsgenie
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/moira-alert/moira"
+	"github.com/moira-alert/moira/senders"
+	"github.com/opsgenie/opsgenie-go-sdk-v2/alert"
+	blackfriday "gopkg.in/russross/blackfriday.v2"
+)
+
+const (
+	titleLimit = 130
+	msgLimit   = 15000
+)
+
+// SendEvents sends the events as an alert to opsgenie
+func (sender *Sender) SendEvents(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) error {
+	createAlertRequest := sender.makeCreateAlertRequest(events, contact, trigger, plot, throttled)
+	_, err := sender.client.Create(context.Background(), createAlertRequest)
+	if err != nil {
+		return fmt.Errorf("error while creating alert: %s", err)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to send %s event message to opsgenie: %s", trigger.ID, err.Error())
+	}
+	return nil
+}
+
+func (sender *Sender) makeCreateAlertRequest(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) *alert.CreateAlertRequest {
+	createAlertRequest := &alert.CreateAlertRequest{
+		Message:     sender.buildTitle(events, trigger),
+		Description: sender.buildMessage(events, throttled, trigger),
+		Responders: []alert.Responder{
+			{Type: alert.EscalationResponder, Name: contact.Value},
+		},
+		Tags:     trigger.Tags,
+		Source:   "Moira",
+		Priority: sender.getMessagePriority(events),
+	}
+
+	if len(plot) > 0 && sender.imageStoreConfigured {
+		imageLink, err := sender.imageStore.StoreImage(plot)
+		if err != nil {
+			sender.logger.Warningf("could not store the plot image in the image store: %s", err)
+		} else {
+			createAlertRequest.Details = map[string]string{
+				"image_url": imageLink,
+			}
+		}
+	}
+
+	return createAlertRequest
+}
+
+func (sender *Sender) buildMessage(events moira.NotificationEvents, throttled bool, trigger moira.TriggerData) string {
+	var message strings.Builder
+
+	desc := trigger.Desc
+	htmlDesc := string(blackfriday.Run([]byte(desc)))
+	htmlDescLen := len([]rune(htmlDesc))
+	charsForHTMLTags := htmlDescLen - len([]rune(desc))
+
+	eventsString := sender.buildEventsString(events, -1, throttled)
+	eventsStringLen := len([]rune(eventsString))
+
+	descNewLen, eventsNewLen := senders.CalculateMessagePartsLength(msgLimit, htmlDescLen, eventsStringLen)
+
+	if htmlDescLen != descNewLen {
+		desc = desc[:descNewLen-charsForHTMLTags] + "...\n"
```

Comment:
I don't understand this logic =( 
Can you explain to me, please? =)
I really don't understand why you additionally cut desc to charsForHTMLTags

Reply:
Because we need to leave space for the html tags. After trimming the desc based on descNewLen we are again converting MD to html using blackfriday. This will add more chars to the desc. charsForHTMLTags is a rough estimate of how many chars will be required for the tags. So we trim the desc to descNewLen - charsForHTMLTags so that when the html tags are added the final len will be less than descNewLen. Of course some chars are also lost in the html conversion because the MD syntax is stripped and replaced with HTML tags. But this conversion cannot be precisely calculated. charsForHTMLTags gives us a safe figure for the chars that might be required.